programs/starter/starter.8
.\"Generated by db2man.xsl. Don't modify this, modify the source.
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "IPSEC_STARTER" 8 "29 Nov 2004" "" ""
.SH NAME
ipsec starter \- start up the IPsec keying daemon (pluto) and load configuration
.SH "SYNOPSIS"
.ad l
.hy 0
.HP 6
\fBipsec\fR \fIstarter\fR [\-\-debug\ \-\-auto_reload\ \fIseconds\fR]
.ad
.hy

.SH "OBSOLETE"

.PP
Note that starter is being obsoleted for the new connection loading code that replace all the scripts, and will be removed from Openswan in the next major release\&.

.SH "DESCRIPTION"

.PP
Openswan Starter is aimed to replace all the scripts which are used to start and stop Openswan, and to do that in a quicker and a smarter way\&.

.PP
It can also reload the configuration file if given a HUP\fI signal, and apply the changes\&.\fR

.PP
What it will do:

.PP
Load and unload KLIPS, or NETKEY (ipsec kernel module)

.PP
Launch and monitor pluto\&.

.PP
Add, initiate, route and delete connections

.PP
Attach and detach interfaces according to config file

.PP
kill \-HUP can be used to reload the config file\&. New connections will be added, old ones will be removed and modified ones will be reloaded\&. Interfaces/Klips/Pluto will be reloaded if necessary\&.

.PP
Upon startup, starter will save its pid to the file /var/run/pluto/ipsec\-starter\&.pid

.PP
Upon reloading, dynamic DNS addresses will be resolved and updated\&. Use \-\-auto_reload to periodicaly check for dynamic DNS changes\&.

.PP
kill \-USR1 can be used to reload all connections\&. This does a \fIdelete\fR, followed by an \fIadd\fR and then either a \fIroute\fR or \fIinitiate\fR operation\&.

.PP
/var/run/pluto/dynip/xxxx can be used to use a virtual interface name in ipsec\&.conf\&. By example, when adsl can be ppp0, ppp1, or some such, one can do:

.PP
\fIipsec\&.conf: interfaces="ipsec0=adsl\fR And use /etc/ppp/ip\-up to create /var/run/pluto/dynip/adsl /var/run/pluto/dynip/adsl: IP_PHYS=ppp0

.PP
%auto can be used to automaticaly name the connections

.PP
kill \-TERM can be used to stop Openswan\&. Pluto will be stopped and kernel modules unloaded\&.

.SH "FILES"

.PP
/etc/ipsec\&.conf

.SH "SEE ALSO"

.PP
\fBipsec\fR(8), \fBipsec_tncfg\fR(8), \fBipsec_pluto\fR(8)

.SH "HISTORY"

.PP
Original by mlafon@arkoon\&.net for Arkoon Network Security\&. Updated for FreeS/WAN version 2 by Michael Richardson <mcr@sandelman\&.ottawa\&.on\&.ca>\&. Merged into Openswan 2\&.2 by Xelerance Corporation

.SH "TODO/BUGS"

.PP
handle wildcards in include lines \-\- use glob() fct ex: include /etc/ipsec\&.*\&.conf

.PP
handle duplicates keywords and sections

.PP
Support also keyword

.PP
add unsupported keywords

.PP
manually keyed connections

.PP
%defaultroute

.PP
IPv6



CHANGES
Please see docs/KNOWN_BUGS file for more information

#testing
v2.5.04
    zero peer_ca to avoid crashes.
    (include glob's may still not work correctly)

v2.5.03
    ipsec.conf parsing should ignore keywords that start with x-
	they are a form of structured comment.
    added in forceencaps= keyword.
    process wildcards with glob() in include statements.

v2.5.02
    fixed bug in ipsec.conf parser, where it could not read values
       that had = in them (such as base64 encoded keys)
    fixed bug in key continuation code that could cause a crash
	if the DNS request timed out after the state was deleted
	for other reasons.
 
v2.5.01
    merged xauthusername code base + multinet tests in.
    removed /dev/hw_random from list of valid random sources on linux.
        use "rngd" instead to feed /dev/random.

v2.5.00 
    fixed various bugs with lifetime values in ipsec.conf parser
    AES-128 (group 5, MD5 or SHA1 for PRF) is now accepted for phase 1,
	and it is now the preferred cipher as well. This should be the
	case for Main mode, Aggressive mode, and for XAUTH client and
	XAUTH server, and PSK and RSA sig mode.

    fixes so that starter will now compile
    move whacklib to lib/libwhack
    adjust makefiles to work with OBJDIR version of Makefile.program
    switch to OBJDIR in programs/* and lib/* if it is defined

    added sanitizer for PID files
    be smarter about including git version info into version
    We need to use /dev/urandom first, as it has more random than /dev/random.
       Otherwise, we run out really fast (within a few minutes)
    Use endian.h when comiling out-of-kernel
    patch to turn error about 17/500,0/0 vs 17/0 error with Cisco VPN3000
       into a warning.
    remove test for NAT-T VID vs NATD payload test. It fails for reasons
       that are unknown at this time, and this check is really being pedantic.
    MAJOR: use starter code for "addconn"
    MAJOR: always use OBJDIRs, and compile on Windows (no kernel)
    MAJOR: includes "Taproom" code --- TCL call outs from pluto at IKE
		transition states. 

v2.4.5
* Fix for compiling on 2.6.14 kernels 
* Refactored natd_lookup / hash code, probably fixes lot of NAT related bugs
  #401 l2tp connection is not work with 2.6 build in IPSEC
  #442 Pluto uses wrong port in NAT-D calculation
  #450 macosx (possible generic PSK+NAT-T rekey bug: eroute already in use.
  #462 updated patch for Openswan and OS X with NAT-T
  #509 KLIPS compilation fail with kernel-2.6.14.2 


v2.4.4
  #487 ASSERTION FAILED at state.c:120:IS_ISAKMP_ENCRYPTED(isakmp_sa->st_state) 
  (see http://www.openswan.org/niscc2/)
  (proper fix in pluto_constants.h)
* Fix for kernels having strstr
* Various gcc4 warning fixes
* disable CONFIG_IPSEC_NAT_TRAVERSAL per default so we can build KLIPS on
  Fedora systems.
* questionable spin_unlock commented out. Might fix reported SMP crashers.
* update to permit alg code without module support
* Fix for detecting proper kernel source/header directory on fedora
* Various bugfixes as reported on http://bugs.openswan.org/
  #499: check for module support in kernel for IPsec Modular Extensions
  #500: recent awk breaks on 'setdefault' command


v2.4.3
  #487 ASSERTION FAILED at state.c:120:IS_ISAKMP_ENCRYPTED(isakmp_sa->st_state) 
  (see http://www.openswan.org/niscc2/)
  (incorrect fixed. version not released)

v2.4.2
* Fixes for compiling on 2.6.14 by David McCullough
* Minor fixes to accomodate FC4 2.6.11 kernels.
* Fix for compilation of KLIPS on 2.4.x kernels.
* Fix for NAT-T on 2.4.31
* Fix for 'short' packets with KLIPS on 2.4.x
* Merged in Jacco's l2tp configuration examples
* Various bugfixes as reported on http://bugs.openswan.org/
  #286 Incorrect links in intro.html
  #344 netkey-acquire patch
  #376 install_ipsec_sa and install_inbound_ipsec_sa
  #486 ASSERTION FAILED at crypto.c:258: key_size==(DES_CBC_BLOCK_SIZE * 3)
  (see http://www.openswan.org/niscc2/)

v2.4.1
* Not publically released

v2.4.0
* NAT-T support for KLIPS on 2.6 (Sponsored by Astaro)
* Additional Cipher support with KLIPS on 2.6 (Sponsored by Astaro)
* Fix for NAT-T/PSK rekey (Ulrich @ Astaro)
* Various bugfixes: #269, #328, #342, #350, #355, #357, #361, #363

v2.3.1
* NAT-T RFC support (mlafon/mcr)
* NAT-T Server Side rewrite - handles rekeying alot better
* NAT-T Client Side rekey bug fixed
* Removed HowTo (obselete)
* IPKG packaging updates
* Log message updates
* dpdaction=restart support 

v2.3.0
* KLIPS for 2.6 support (Experimental)
  [ good results on FC3-AMD and vanilla/debian kernel source, but not
    FC3-intel. Might be the grsecurity patch  ]
* Aggressive Mode Support (client and server)
* IKE Mode Config support (Experimental)
* Cisco VPN 3xxx client Interop (Experimental)
* Cryptographic helpers framework
* Fixes for NAT-T on 2.4.28+ kernels.

v2.2.0
* Added RFC 3706 DPD support (see README.DPD)
* Added AES from JuanJo's ALG patches
* Fixes for /proc filesystem issues that started to appear in 2.4.25

v2.1.2
* Fix loading of 2.6 modules 
* Fix for snprintfs() in /proc, new for 2.4.25 kernels (dhr/pw)
* Fix checks for some log files/dirs in case they are sockets or pipes (pw)
* Fix for crl.pem crash/core (dhr/as/kb)

v2.1.1
* Fix _pluto_adns installation path (kb)
* Fix sending of X.509 CR's when no CA present (mcr)

v2.1.0
* NAT-T support (Mathieu Lafon - Arkoon)
* X.509 fixes (Andreas Steffan)
* New configuration file directive, {left|right}sourceip=#.#.#.# 
  This will set the source address when talking to a particular 
  connection.  This is very usefull to assign a static IP to your laptop 
  while travelling.  This is based on Tuomo Soini's Advanced Routing 
  patch.

RCSID $Id: CHANGES,v 1.326.2.31 2005/11/24 05:35:43 ken Exp $


programs/starter/starter.c
/* Openswan IPsec starter (starter.c)
 * Copyright (C) 2001-2002 Mathieu Lafon - Arkoon Network Security
 * Copyright (C) 2004 Xelerance Corporation
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.  See <http://www.fsf.org/copyleft/gpl.txt>.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 *
 * RCSID $Id: starter.c,v 1.12 2005/01/11 17:52:51 ken Exp $
 */

#include <sys/types.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <sys/queue.h>
#include <stdlib.h>
#include <stdio.h>
#include <signal.h>
#include <unistd.h>
#include <sys/time.h>
#include <time.h>
#include <string.h>
#include <errno.h>
#include <fcntl.h>

<<<<<<< master
#include "oswlog.h"
#include "ipsecconf/confread.h"
#include "ipsecconf/confwrite.h"
#include "ipsecconf/starterlog.h"
#include "ipsecconf/files.h"
#include "ipsecconf/starterwhack.h"
#include "ipsecconf/pluto.h"
#include "ipsecconf/klips.h"
#include "ipsecconf/netkey.h"
#include "ipsecconf/cmp.h"
#include "ipsecconf/interfaces.h"
#include "ipsecconf/keywords.h"
=======
#include "oswlog.h"
#include "confread.h"
#include "confwrite.h"
#include "starterlog.h"
#include "files.h"
#include "starterwhack.h"
#include "pluto.h"
#include "klips.h"
#include "netkey.h"
#include "cmp.h"
#include "interfaces.h"
#include "keywords.h"
>>>>>>> public

#define FLAG_ACTION_START_PLUTO   0x01
#define FLAG_ACTION_RELOAD        0x02
#define FLAG_ACTION_FORCE_RELOAD  0x04
#define FLAG_ACTION_QUIT          0x08
#define FLAG_ACTION_LISTEN        0x10

static unsigned int _action_ = 0;
int verbose;
int warningsarefatal = 0;

char configfile[PATH_MAX];
char rootdir[PATH_MAX];       /* when evaluating paths, prefix this to them */
int showonly = 0;
char *progname;

static void fsig(int signal)
{
	 switch (signal) {
	 	case SIGCHLD: {
				int status;
				pid_t pid;
				char *name = NULL;

				while ((pid=waitpid(-1,&status,WNOHANG))>0) {
					if (pid == starter_pluto_pid()) name = " (Pluto)";
					if (WIFSIGNALED(status))
						starter_log(LOG_LEVEL_INFO, 
							"child %d%s has been killed by sig %d\n",
							pid, name?name:"", WTERMSIG(status));
					else if (WIFSTOPPED(status))
						starter_log(LOG_LEVEL_INFO, 
							"child %d%s has been stopped by sig %d\n",
							pid, name?name:"", WSTOPSIG(status));
					else if (WIFEXITED(status))
						starter_log(LOG_LEVEL_INFO, 
							"child %d%s has quit (exit code %d)\n",
							pid, name?name:"", WEXITSTATUS(status));
					else
						starter_log(LOG_LEVEL_INFO, 
							"child %d%s has quit", pid, name?name:"");

					if (pid == starter_pluto_pid())
						starter_pluto_sigchild(pid);
				}
			}
			break;

		case SIGPIPE:
			/** ignore **/
			break;

		case SIGALRM:
			_action_ |= FLAG_ACTION_START_PLUTO;
			break;

		case SIGHUP:
			_action_ |= FLAG_ACTION_RELOAD;
			break;

		case SIGTERM:
		case SIGQUIT:
		case SIGINT:
			_action_ |= FLAG_ACTION_QUIT;
			break;

		case SIGUSR1:
			_action_ |= FLAG_ACTION_FORCE_RELOAD;
			_action_ |= FLAG_ACTION_RELOAD;
			break;

		default:
			starter_log(LOG_LEVEL_ERR, 
				"fsig(): unknown signal %d -- investigate", signal);
			break;
	 }
}

static void usage(char *name)
{
	fprintf(stderr, "Usage: %s [--debug] [--auto_reload <x sec>]\n", name);
	exit(1);
}

void exit_tool(num)
{
    exit(num);
}

int main (int argc, char **argv)
{
	struct starter_config *cfg = NULL, *new_cfg;
	struct starter_conn *conn, *conn2;
	struct stat stb;
	char *err = NULL;
	int i, debug=0, no_fork=0;
	struct timeval tv;
	unsigned long auto_reload = 0;
	time_t last_reload;
	int id = 1;
	extern int yydebug;
	char *confdir;
	bool justdump = FALSE;

	progname=argv[0];

	/* find environment location for /etc */
	confdir = getenv(IPSEC_CONFDIR_VAR);
	if(confdir == NULL)
	{
	    confdir = IPSEC_CONFDIR;
	}
		
	/* calculate default value for configfile */
	configfile[0]='\0';
	strncat(configfile, confdir, sizeof(configfile));
	if(configfile[strlen(configfile)-1]!='/')
	{
	    strncat(configfile, "/", sizeof(configfile));
	}
	strncat(configfile, "ipsec.conf", sizeof(configfile));

	/**
	 * Parse command line
	 */
	for (i=1; i<argc; i++) {
		if (strcmp(argv[i],"--debug")==0) {
			debug = 1;
			no_fork = 1;
		}
		else if (strcmp(argv[i],"--no_fork")==0) {
			no_fork = 1;
		}
		else if (strcmp(argv[i],"--configfile")==0 && (i+1)<argc && argv[i+1]!=NULL) {
		    strcpy(configfile, argv[i+1]);
		    i++;
		}
		else if (strcmp(argv[i],"--rootdir")==0 && (i+1)<argc && argv[i+1]!=NULL) {
		    strcpy(rootdir, argv[i+1]);
		    i++;
		}
		else if (strcmp(argv[i],"--parsedebug")==0) {
  	                yydebug = 1;
		}
		else if (strcmp(argv[i],"--verbose")==0) {
		    verbose = 1;
		}
		else if (strcmp(argv[i],"-Werror")==0) {
		    warningsarefatal = 1;
		}
		else if (strcmp(argv[i],"--dumpcfg")==0) {
		    justdump = TRUE;
		}
		else if (strcmp(argv[i],"--showonly")==0) {
			showonly = 1;
			warningsarefatal = 0;
		}
		else if ((strcmp(argv[i],"--auto_reload")==0) && (i+1 < argc)) {
			auto_reload = atoi(argv[++i]);
			if (!auto_reload) usage(argv[0]);
		}
		else {
			usage(argv[0]);
		}
	}

	starter_use_log (debug, 1, debug ? 0 : 1);

	/**
	 * Init
	 */
	signal(SIGHUP,fsig);
	signal(SIGCHLD,fsig);
	signal(SIGPIPE,fsig);
	signal(SIGINT,fsig);
	signal(SIGTERM,fsig);
	signal(SIGQUIT,fsig);
	signal(SIGALRM,fsig);
	signal(SIGUSR1,fsig);

	/**
	 * Verify that we can start
	 */
	if (!showonly && getuid()!=0) {
		starter_log(LOG_LEVEL_ERR, "ERROR: Must be root (uid=0)");
		exit(1);
	}

	if (!showonly && stat(PID_FILE,&stb)==0) {
		starter_log(LOG_LEVEL_ERR,
			"ERROR: pluto is already running (%s exists) -- aborting", PID_FILE);
		exit(1);
	}

	if (stat(DEV_RANDOM,&stb)!=0) {
		starter_log(LOG_LEVEL_ERR, "ERROR: Unable to start Openswan IPsec, no %s!",
			DEV_RANDOM);
		exit(1);
	}

	if (stat(DEV_URANDOM,&stb)!=0) {
		starter_log(LOG_LEVEL_ERR, "ERROR: Unable to start Openswan IPsec, no %s!",
			DEV_URANDOM);
		exit(1);
	}

	cfg = confread_load(configfile, &err);
	if (!cfg) {
		starter_log(LOG_LEVEL_ERR, "ERROR: Can't load config: %s", err ? err : "unknown error");
		if (err) free(err);
		exit(1);
	}
	if (err) free(err);

	if(justdump)
	{
	    confwrite(cfg,stdout);
	    exit(0);
	}

	/* Need to determine which stack to use here, and if() it */
	if ((starter_klips_init()!=0) || (starter_klips_set_config(cfg)!=0)) {
		exit(1);
	}

	starter_ifaces_init();
	starter_ifaces_clear();

	last_reload = time(NULL);

	starter_log(LOG_LEVEL_INFO, "Starting Openswan IPsec %s [starter]...",
		ipsec_version_code());

	/**
	 * Fork if we're not debugging stuff
	 */
	if (!no_fork) {
		switch (fork()) {
			case 0:
				starter_use_log (0, 0, 1);
				{
					int fnull;
					fnull = open("/dev/null", O_RDWR);
					if (fnull >= 0) {
						dup2(fnull,STDIN_FILENO);
						dup2(fnull,STDOUT_FILENO);
						dup2(fnull,STDERR_FILENO);
						close(fnull);
					}
				}
				break;
			case -1:
				starter_log(LOG_LEVEL_ERR, "ERROR: Can't fork: %s", strerror(errno));
				break;
			default:
				exit(0);
				break;
		}
	}

	/**
	 * Save pid file in /var/run/pluto/starter.pid
	 */
	{
		FILE *f = fopen(MY_PID_FILE, "w");
		if (f) {
			fprintf(f, "%u\n", getpid());
			fclose(f);
		}
	}

	starter_ifaces_load (cfg->setup.interfaces, cfg->setup.options[KBF_OVERRIDEMTU],
#ifdef NAT_TRAVERSAL
		cfg->setup.nat_traversal ? 1 : 0
#else
		0
#endif
		);

	_action_ = FLAG_ACTION_START_PLUTO;

	for (;;) {

		/**
		 * Stop pluto (if started) and exit
		 */
		if (_action_ & FLAG_ACTION_QUIT) {
			if (starter_pluto_pid()) {
				starter_stop_pluto();
			}
			starter_ifaces_clear();
			/* Need to determine which stack here... */
			starter_klips_cleanup();
			starter_netkey_cleanup();

			confread_free(cfg);
			starter_log(LOG_LEVEL_DEBUG, "ipsec starter stopped");
			unlink(MY_PID_FILE);
			exit(0);
		}

		/**
		 * Delete all connections. Will be added below
		 */
		if (_action_ & FLAG_ACTION_FORCE_RELOAD) {
			if (starter_pluto_pid()) {
			    for(conn = cfg->conns.tqh_first; conn != NULL; conn = conn->link.tqe_next)
			    {
				if (conn->state == STATE_ADDED) {
				    starter_whack_del_conn(conn);
				    conn->state = STATE_TO_ADD;
				}
			    }
			}
			_action_ &= ~FLAG_ACTION_FORCE_RELOAD;
		}

		/**
		 * Reload a new config gile
		 */
		if (_action_ & FLAG_ACTION_RELOAD) {
			err = NULL;
			starter_log(LOG_LEVEL_INFO, "Reloading config %s...", configfile);
			new_cfg = confread_load(configfile, &err);
			if (new_cfg) {
				/**
				 * Switch to new config. New conn will be loaded below
				 */
				/* Check stacks here too */
				if (starter_cmp_klips(cfg, new_cfg)) {
					starter_log(LOG_LEVEL_DEBUG, "Klips has changed");
					starter_klips_set_config(new_cfg);
				}

				if (starter_ifaces_load (new_cfg->setup.interfaces,
					new_cfg->setup.options[KBF_OVERRIDEMTU],
#ifdef NAT_TRAVERSAL
					new_cfg->setup.nat_traversal ? 1 : 0
#else
					0
#endif
					)) {
					_action_ |= FLAG_ACTION_LISTEN;
				}

				if (starter_cmp_pluto(cfg, new_cfg)) {
					starter_log(LOG_LEVEL_DEBUG, "Pluto has changed");
					if (starter_pluto_pid()) {
						starter_stop_pluto();
					}
					_action_ &= ~FLAG_ACTION_LISTEN;
					_action_ |= FLAG_ACTION_START_PLUTO;
				}
				else {
				    /**
				     * Only reload conns if pluto is not killed
				     */
				    /**
				     * Look for new connections that are already loaded
				     */
				    for(conn = cfg->conns.tqh_first;
					conn != NULL;
					conn = conn->link.tqe_next)
				    {
					if (conn->state == STATE_ADDED)
					{
					    for(conn2 = cfg->conns.tqh_first; conn2 != NULL; conn2 = conn2->link.tqe_next)
					    {
						if ((conn2->state == STATE_TO_ADD) &&
						    (starter_cmp_conn(conn,conn2)==0)) {
						    conn->state = STATE_REPLACED;
						    conn2->state = STATE_ADDED;
						    conn2->id = conn->id;
						    break;
						}
					    }
					}
				    }
				    /**
				     * Remove now unused conn
				     */
				    for(conn = cfg->conns.tqh_first; conn != NULL; conn = conn->link.tqe_next)
				    {
					if (conn->state == STATE_ADDED)
					{
					    starter_whack_del_conn(conn);
					}
				    }
				}
				confread_free(cfg);
				cfg = new_cfg;
			}
			else {
				starter_log(LOG_LEVEL_ERR,
					"can't reload config file: %s -- keeping old one", err);
			}
			if (err) free(err);
			_action_ &= ~FLAG_ACTION_RELOAD;
			last_reload = time(NULL);
		}

		/**
		 * Start pluto
		 */
		if (_action_ & FLAG_ACTION_START_PLUTO) {
			if (starter_pluto_pid()==0) {
				starter_log(LOG_LEVEL_INFO, "Attempting to start pluto...");
				/* Again, check stacks here and if() */
				starter_klips_clear();
				starter_netkey_clear();
				if (starter_start_pluto(cfg,debug)==0) {
					starter_whack_listen();
				}
				else {
					/** schedule next try **/
					alarm(PLUTO_RESTART_DELAY);
				}
			}
			_action_ &= ~FLAG_ACTION_START_PLUTO;
  		        for(conn = cfg->conns.tqh_first; conn != NULL; conn = conn->link.tqe_next)
			{
			    if (conn->state == STATE_ADDED)
				conn->state = STATE_TO_ADD;
			}
		}

		/**
		 * Tell pluto to reread its interfaces
		 */
		if (_action_ & FLAG_ACTION_LISTEN) {
			starter_whack_listen();
			_action_ &= ~FLAG_ACTION_LISTEN;
		}

		/**
		 * Add stale connections
		 */
		if (starter_pluto_pid()!=0) {
		    for(conn = cfg->conns.tqh_first; conn != NULL; conn = conn->link.tqe_next)
		    {
			switch(conn->desired_state)
			{
			case STARTUP_NO:
			    break;

			case STARTUP_POLICY:
			    /* mark conn as not negotiable
			     * starter_whack_route_conn(conn);
			     */
			    break;
			    
			case STARTUP_ADD:
			case STARTUP_START:
			case STARTUP_ROUTE:
			    if (conn->state == STATE_LOADED)
			    {
				if (conn->id == 0) {
				    /**
				     * Affect new unique id
				     */
				    conn->id = id++;
				}
				starter_whack_add_conn(conn);
				conn->state = STATE_ADDED;
			    }
			    break;
			}
			
			if(conn->state == STATE_ADDED)
			{
			    switch(conn->desired_state)
			    {
			    case STARTUP_NO:
			    case STARTUP_POLICY:
			    case STARTUP_ADD:
				break;
				
			    case STARTUP_START:
				starter_whack_initiate_conn(conn);
				break;
				
			    case STARTUP_ROUTE:
				starter_whack_route_conn(conn);
				break;
			    }
			    
			    conn->state = STATE_UP;
			}
		    }
		}
		
		/**
		 * If auto_reload activated, when to stop select
		 */
		if (auto_reload) {
		    time_t now = time(NULL);
		    tv.tv_sec = (now<last_reload+auto_reload) ?
			(last_reload+auto_reload-now) : 0;
		    tv.tv_usec = 0;
		}
		
		/**
		 * Wait for something to happend
		 */
		if (select (0, NULL, NULL, NULL, auto_reload ? &tv : NULL) == 0) {
		    /**
		     * Timeout -> auto_reload
		     */
		    _action_ |= FLAG_ACTION_RELOAD;
		}
}

	return 0;
}

/*
 * Local Variables:
 * c-basic-offset:4
 * c-style: pluto
 * End:
 */
